### PR TITLE
Feature/dataset latest metadata

### PR DESCRIPTION
**Pull request by jamie-short**
added in command to get:
- dataset --id {dataset_id} --version-id {version_id} [--long]

The command takes an optional flag for long, short is default.
- Outputs meta data & a table of the associated datafiles with name, size & format

additional data outputted if long passed:
- includes all additional fields shown on current dataset page in UI

**Note**
I have not refactored the Model tables code to sue the new generic table output code.
This will be addressed on a separate branch to prevent merge conflicts with current Models work

### CODE REVIEW

File: dafni_cli/utils.py [hunk 1]
```diff
@@ -70,4 +72,137 @@ def process_date_filter(date_str: str) -> str:
...
+    if value:
+        try:
+            return isoparse(value).strftime("%B %d %Y")
+        except ValueError:
+            return value
+    return default
+
+
+def output_table_row(
+    columns: List[str], widths: List[int], alignment: str = "<", header: bool = False
```
Comment (iroper-tessella) [inline dafni_cli/utils.py:137]: Missing `header` in docstring

File: dafni_cli/utils.py [hunk 2]
```diff
@@ -70,4 +72,137 @@ def process_date_filter(date_str: str) -> str:
...
+    return default
+
+
+def process_dict_datetime(
+    input_dict: dict, key: str, nested_key: Optional[str] = None, default="N/A"
+) -> str:
+    """Utility function to check a nested dict for a given
+    key and nested key if applicable. If the keys exist, the
+    associated value is parsed to a datetime and then formatted to an
+    applicable datetime string, otherwise the default value is returned.
```
Comment (iroper-tessella) [inline dafni_cli/utils.py:116]: This description doesn't mention that this only processes times that are in an ISO date format. Could this be added?

File: test/test_utils.py [hunk 3]
```diff
@@ -259,3 +259,454 @@ def test_value_error_raised_if_given_invalid_formatted_date(self, date_str):
...
+            == "Column_1  Column_2   \n----------\nValue_1   Value_2    \nValue_3   Value_4    \n"
+        )
+        assert mock_output.call_args_list == [
+            call(columns, widths, "<", header=True),
+            call(rows[0], widths, "<"),
+            call(rows[1], widths, "<"),
+        ]
+
+
+class TestFileSize:
```
Comment (iroper-tessella) [inline test/test_utils.py:677]: Rename to `TestProcessFileSize` for consistency

File: test/test_utils.py [hunk 4]
```diff
@@ -259,3 +259,454 @@ def test_value_error_raised_if_given_invalid_formatted_date(self, date_str):
...
+            (1e7, "10.0 MB"),
+            (1e8, "100.0 MB"),
+            (1e9, "1.0 GB"),
+            (1e10, "10.0 GB"),
+            (1e11, "100.0 GB"),
+            (1e12, "1000.0 GB"),
+            (1.5e3, "1.5 KB"),
+            (2.7e6, "2.7 MB"),
+            (3.8e9, "3.8 GB"),
+        ],
```
Comment (iroper-tessella) [inline test/test_utils.py:706]: Might be worth adding a couple of cases where there are more than 2 significant figures to ensure the result of a rounded number is what we expect.

**Review by iroper-tessella — CHANGES_REQUESTED** [review]
Quite a large PR so quite a few comments. Most a typos/docstring stuff but there are a few things I have suggested changing. If you disagree/I've gotten something wrong, let's discuss.

File: test/test_utils.py (hunk 4 above)
Comment (jamie-short) [reply test/test_utils.py:706]: thats a very good idea actually, good spot

File: dafni_cli/utils.py [hunk 5]
```diff
@@ -70,4 +72,137 @@ def process_date_filter(date_str: str) -> str:
...
+
+    Returns:
+        str: Parsed datetime associated with given dict and key(s)
+    """
+    value = check_key_in_dict(input_dict, key, nested_key, None)
+    if value:
+        try:
+            return isoparse(value).strftime("%B %d %Y")
+        except ValueError:
+            return value
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply dafni_cli/utils.py:132]: it's a tricky one, as you would assume you wouldn't call this function if you weren't expecting a datetime, however this covers the case where the datetime is not in an iso format, which I come across with Toms model for example

File: dafni_cli/utils.py [hunk 6]
```diff
@@ -70,4 +72,137 @@ def process_date_filter(date_str: str) -> str:
...
+    return default
+
+
+def output_table_row(
+    columns: List[str], widths: List[int], alignment: str = "<", header: bool = False
+) -> str:
+    """Function to generate a table header for the given keys,
+    with a dashed line of equal length underneath
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply dafni_cli/utils.py:140]: sounds reasonable

File: dafni_cli/utils.py [hunk 7]
```diff
@@ -70,4 +72,137 @@ def process_date_filter(date_str: str) -> str:
...
+
+    Args:
+        columns (List[str]): Table Columns
+        widths (List[int]): Width associated with each column
+        alignment (str, optional): Direction to align header. Defaults to "<".
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply dafni_cli/utils.py:145]: updated

File: dafni_cli/utils.py [hunk 8]
```diff
@@ -70,4 +72,137 @@ def process_date_filter(date_str: str) -> str:
...
+
+    Returns:
+        str: String containing all Headers with associated spacing and line breaks
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply dafni_cli/utils.py:148]: added

File: test/datasets/test_dataset_metadata.py [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+        def test_set_details_from_dict_called_if_give_dict(self, mock_set):
+            # SETUP
+            dataset_dict = {"key": "value"}
+            # CALL
+            DatasetMeta(dataset_dict)
+            # ASSERT
+            mock_set.assert_called_once_with(dataset_dict)
+
+    class TestSetDetailsFromDict:
+        """Test class to test the set_dateils_from_dict functionality"""
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply test/datasets/test_dataset_metadata.py:153]: updated

File: dafni_cli/utils.py (hunk 1 above)
Comment (jamie-short) [reply dafni_cli/utils.py:137]: updated

File: dafni_cli/utils.py (hunk 2 above)
Comment (jamie-short) [reply dafni_cli/utils.py:116]: updated

File: test/test_utils.py (hunk 3 above)
Comment (jamie-short) [reply test/test_utils.py:677]: updated

File: dafni_cli/utils.py [hunk 10]
```diff
@@ -70,4 +72,137 @@ def process_date_filter(date_str: str) -> str:
...
+    """
+    table_items = [
+        f"{column:{alignment}{widths[idx]}}" for idx, column in enumerate(columns)
+    ]
+    table_str = " ".join(table_items)
+    table_str += "\n"
+
+    if header:
+        table_str += "-" * sum(widths, len(columns))
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply dafni_cli/utils.py:157]: So this was done to cover the case where the contents of the column matches the width. So this adds an extra space in to ensure there is always a gap

File: dafni_cli/utils.py [hunk 11]
```diff
@@ -1,7 +1,9 @@
 import click
 import textwrap
-from typing import List
+from typing import List, Optional, Union
 from datetime import datetime as dt
+from dateutil.parser import isoparse
+from dafni_cli.consts import DATA_FORMATS
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply dafni_cli/utils.py:6]: removed

File: dafni_cli/utils.py [hunk 12]
```diff
@@ -70,4 +72,137 @@ def process_date_filter(date_str: str) -> str:
...
+    table_str = output_table_row(columns, widths, alignment, header=True)
+    rows = [output_table_row(row, widths, alignment) for row in values]
+    table_str += "".join(rows)
+
+    return table_str
+
+
+def process_file_size(file_size: Union[int, float]) -> str:
+    """Utility function to take in a file size in bytes
+    and format into a tale ready format.
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply dafni_cli/utils.py:186]: typo, fixed

File: dafni_cli/datasets/dataset_metadata.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+
+
+class DataFile:
+    """Class to represent the DAFNI Dataset File
+    client model
+
+    Methods:
+        __init__(): DataFile constructor
+
+    Attributes:
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply dafni_cli/datasets/dataset_metadata.py:23]: added

File: dafni_cli/datasets/dataset_metadata.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+        __init__(): DatasetMeta constructor
+        set_details_from_dict(jwt (str), dataset (dict)): Sets the dataset metadata attributes from given client model dict
+        output_metadata_details(): Prints key information of the dataset metadata to console.
+        output_datafiles_table(): Prints a table to the console of all File related information
+        output_metadata_extra_details(): Prints extra details relating to the Dataset Metadata
+
+    Attributes:
+        created: Date the dataset was created
+        creator: Creator of the dataset
+        contact: Conact relating to the dataset
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply dafni_cli/datasets/dataset_metadata.py:72]: fixed

File: dafni_cli/datasets/dataset_metadata.py [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+    Metadata client model
+
+    Methods:
+        __init__(): DatasetMeta constructor
+        set_details_from_dict(jwt (str), dataset (dict)): Sets the dataset metadata attributes from given client model dict
+        output_metadata_details(): Prints key information of the dataset metadata to console.
+        output_datafiles_table(): Prints a table to the console of all File related information
+        output_metadata_extra_details(): Prints extra details relating to the Dataset Metadata
+
+    Attributes:
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply dafni_cli/datasets/dataset_metadata.py:69]: added

File: dafni_cli/datasets/dataset_metadata.py [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+        self.themes = None
+        self.publisher = None
+        self.issued = None
+        self.rights = None
+        self.language = None
+        self.standard = None
+        self.update = None
+
+        if dataset_dict:
+            self.set_details_from_dict(dataset_dict)
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply dafni_cli/datasets/dataset_metadata.py:110]: It makes testing easier

File: dafni_cli/datasets/dataset_metadata.py [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+
+    def output_metadata_details(self, long: bool = False):
+        """Function to output details relating to the Dataset.
+        The default behavior is to print all standard metadata and
+        a table relating to the associated files.
+        If the long option is given, the additional metadata fields are
+        also printed to the console
+
+        Args:
+            long (bool, optional): Flag to print additional metdata. Defaults to False.
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply dafni_cli/datasets/dataset_metadata.py:160]: fixed

File: dafni_cli/datasets/dataset_metadata.py [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+        """
+        click.echo(f"\nCreated: {self.created}")
+        click.echo(f"Creator: {self.creator}")
+        click.echo(f"Contact: {self.contact}")
+        click.echo("Description:")
+        prose_print(self.description, CONSOLE_WIDTH)
+        click.echo("Identifier: ")
+        prose_print(" ".join(self.identifier), CONSOLE_WIDTH)
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply dafni_cli/datasets/dataset_metadata.py:168]: So its a list of id type strings, but some of them are  much wider than the console, such as urls, or asset ids which are 3 guids concatenated, so this forces the output to be a bit more readable

File: dafni_cli/datasets/dataset_metadata.py [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+        self.name = check_key_in_dict(file_dict, "spdx:fileName")
+        self.size = process_file_size(
+            check_key_in_dict(file_dict, "dcat:byteSize", default=None)
+        )
+        self.format = DATA_FORMATS.get(
+            check_key_in_dict(file_dict, "dcat:mediaType"), ""
+        )
+
+
+class DatasetMeta:
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply dafni_cli/datasets/dataset_metadata.py:58]: updated to DatasetMetadata

File: dafni_cli/get.py [hunk 20]
```diff
@@ -27,16 +28,19 @@ def get(ctx: Context):
     "--long/--short",
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply dafni_cli/get.py:28]: sure

File: test/datasets/test_dataset_metadata.py [hunk 21]
```diff
@@ -0,0 +1,413 @@
...
+
+from test.fixtures.dataset_fixtures import (
+    dataset_metadata_fixture,
+    datafile_mock,
+    dataset_meta_mock,
+)
+
+
+class TestDataFile:
+    "Test class to test the DataFile Class"
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply test/datasets/test_dataset_metadata.py:15]: updated

File: test/datasets/test_dataset_metadata.py [hunk 22]
```diff
@@ -0,0 +1,413 @@
...
+            file_dict = {"key": "value"}
+            # CALL
+            DataFile(file_dict)
+            # ASSERT
+            mock_set.assert_called_once_with(file_dict)
+
+    @patch("dafni_cli.datasets.dataset_metadata.process_file_size")
+    @patch("dafni_cli.datasets.dataset_metadata.check_key_in_dict")
+    class TestSetDetailsFromDict:
+        """Test class to test the DataFile.init functionality"""
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply test/datasets/test_dataset_metadata.py:40]: updated

File: test/datasets/test_dataset_metadata.py [hunk 23]
```diff
@@ -0,0 +1,413 @@
...
+
+            # ASSERT
+            assert instance.format == expected
+
+
+class TestDatasetMeta:
+    """Test class for the DatasetMeta class"""
+
+    class TestInit:
+        "Test class for the DatasetMeta.__init__() function"
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply test/datasets/test_dataset_metadata.py:113]: updated

File: test/datasets/test_dataset_metadata.py [hunk 24]
```diff
@@ -0,0 +1,413 @@
...
+
+        def test_datafile_has_expected_attributes(self, mock_set):
+            # SETUP
+            expected_attr = ["name", "size", "format"]
+            # CALL
+            instance = DataFile()
+            # ASSERT
+            assert all(getattr(instance, attr) is None for attr in expected_attr)
+
+        def test_set_Details_from_dict_called_if_given_a_dict(self, mock_set):
```
[the remark this answers is not in the document]
Comment (jamie-short) [reply test/datasets/test_dataset_metadata.py]: updated

File: dafni_cli/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (iroper-tessella) [reply dafni_cli/utils.py:146]: I don't think this was added.

File: dafni_cli/utils.py [hunk 25]
```diff
@@ -86,29 +85,27 @@ def check_key_in_dict(
 
     Args:
         input_dict (dict): dict to check in for keys
-        key (str): first key to check for
-        nested_key (Optional[str], optional): A nested key to check for. Defaults to None.
+        key (str): keys to check for
```
Comment (iroper-tessella) [inline dafni_cli/utils.py:88]: This isn't quite right. Should be `keys (List[str]): keys to check for`

**Review by iroper-tessella — CHANGES_REQUESTED** [review]
Just a couple of things that have been missed I think.

**Review by iroper-tessella — APPROVED** [review]
All requested changes made, happy to be merged.

**Pull request merged**